server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RecipientRewriteTableProcessorTest.java

```diff
@@ -0,0 +1,344 @@
+        when(virtualTableStore.getMappings(any(String.class), any(String.class))).thenReturn(null);
+
+        mail = FakeMail.builder()
+            .mimeMessage(message)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .build();
+
+        processor.processMail(mail, mockMailetContext);
+
+        assertThat(mail.getRecipients()).containsOnly(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void processShouldWriteLogAndSendMailToAllErrorRecipientsWhenErrorMappingException() throws Exception {
+        when(virtualTableStore.getMappings(eq("other"), eq(MailAddressFixture.JAMES_APACHE_ORG))).thenThrow(ErrorMappingException.class);
+
+        mail = FakeMail.builder()
+            .mimeMessage(message)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .build();
+
+        processor.processMail(mail, mockMailetContext);
+
+        verify(mockMailetContext, times(1)).log(eq(LogLevel.INFO), eq("Error while process mail."), any(MessagingException.class));
+        verify(mockMailetContext).sendMail(eq(mail.getSender()), eq(ImmutableList.of(MailAddressFixture.OTHER_AT_JAMES)), eq(message), eq(Mail.ERROR));
+        verifyNoMoreInteractions(mockMailetContext);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void processShouldWriteLogAndSendMailToAllErrorRecipientsWhenRecipientRewriteTableException() throws Exception {
+        when(virtualTableStore.getMappings(eq("other"), eq(MailAddressFixture.JAMES_APACHE_ORG))).thenThrow(RecipientRewriteTableException.class);
+
+        mail = FakeMail.builder()
+            .mimeMessage(message)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .build();
+
+        processor.processMail(mail, mockMailetContext);
+
+        verify(mockMailetContext, times(1)).log(eq(LogLevel.INFO), eq("Error while process mail."), any(MessagingException.class));
+        verify(mockMailetContext).sendMail(eq(mail.getSender()), eq(ImmutableList.of(MailAddressFixture.OTHER_AT_JAMES)), eq(message), eq(Mail.ERROR));
+        verifyNoMoreInteractions(mockMailetContext);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void processShouldWriteLogAndSendMailToAllErrorRecipientsWhenMessagingException() throws Exception {
+        when(virtualTableStore.getMappings(eq("other"), eq(MailAddressFixture.JAMES_APACHE_ORG))).thenThrow(MessagingException.class);
+
+        mail = FakeMail.builder()
+            .mimeMessage(message)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .build();
+
+        processor.processMail(mail, mockMailetContext);
+
+        verify(mockMailetContext, times(1)).log(eq(LogLevel.INFO), eq("Error while process mail."), any(MessagingException.class));
```

Comment:
Use FakeMailetContext instead.

It offers a getSentMail on which you can assert the mail that were sent. This way we do not need mocking, and hence are less bound to the implementation.